src/rebar_compiler_dag.erl

```diff
@@ -269,10 +269,53 @@ compile_order(G, AppDefs, SrcExt, ArtifactExt) ->
                     end
             end
         end, new_cache(), digraph:edges(G)),
-    Sorted = lists:reverse(digraph_utils:topsort(AppDAG)),
+    Standalone = [Name || {Name, _} <- AppDefs],
+    Sorted = interleave(Standalone, AppDAG),
     digraph:delete(AppDAG),
-    Standalone = [Name || {Name, _} <- AppDefs] -- Sorted,
-    Standalone ++ Sorted.
+    Sorted.
+
+%% Assume that the standalone app list respects the
+%% rebar.config deps order, and enforce the sorted app
+%% constraints onto it such that we're always respecting
+%% the hard dependencies.
+%%
+%% What we do here is a sort of run-length reordering based
+%% on DAG information, which preserves the original dependency
+%% order as declared, but successfully interleaves hard deps
+%% to come first.
+%%
+%% Note that this approach is required as opposed to topsort
+%% because when the original DAG reports two distinct set of
+%% app dependencies that are joined by an invisible compile-time
+%% one (e.g. a parse_transform runtime dep between both sets),
+%% then the topological sort can't provide the right ordering
+%% information because it's flattened into one list, but
+%% this one can.
+interleave(Apps, DAG) ->
+     interleave(Apps, DAG, #{}).
+
+interleave([], _, _) ->
+    [];
+interleave([App|Apps], DAG, Expanded) ->
+    case Expanded of
+        #{App := _} ->
```

Comment:
Could you be using the guard `is_map_key` in `interleave` and `dedupe` instead?

Reply:
Actually, why is a map used, is `Expanded` not just acting as a `set`?

Reply:
it's acting as a set, but the map is generally faster than the set. I can swap the datastructure if you want though.

Reply:
Ah, ok, I'm fine either way. I did recall discussion about switching some of these data structures internally to use maps and that lead me to seeing `sets` has a "version 2", https://erlang.org/doc/man/sets.html#new-1

Since we aren't doing any `set` operations besides `is_element` I don't think it is too important to switch, but seems we can in the future and possibly not lose speed.

Reply:
Yeah I'm good going for the semantics and using the sets module then. Will update later today.